### PR TITLE
Symbol outline / search: Customizable display text

### LINKED ISSUE #220: Symbol outline / search: Customizable display text
It would be nice to have the option to customize how symbol outline / search will be displayed.
This could be done by using the attribute on the displayed element.

```xml
<Element1>
  <Element2 id="id1" />
   <Element2 id="id2" />
   <Element2 id="id3" />
</Element1>
```

In my case my symbol outline looks like this:
```
Element1
--Element2
--Element2
--Element2
````

It would be nice, if it could look sth like this:
```
Element1
--Element2 (id1)
--Element2 (id2)
--Element2 (id3)
````

Thx. :)

### PR DESCRIPTION
**Pull request by angelozerr**
Symbol outline / search: Customizable display text

Fixes #220

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Review by rgrunber — COMMENTED** [review]
Just some quick comments.

File: docs/Symbols.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# Symbols
+
+## Symbols filter settings
+
+By default Outline (symbols) display `DOM elements`, `processing instruction` and `DTD element, entity, attribute list declarations`.
+
+The DOM attributes and text nodes are not displayed in the outline. Indeed to prevent from good performance, symbols must be limitated; displaying attributes and text nodes in the outline could generate a lot of symbols.
```
Comment (rgrunber) [inline docs/Symbols.md:7]: Indeed to ensure good performance, symbols must be limited;

File: docs/Symbols.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+![Outline of pom.xml with text filter](images/Symbols/SymbolsPOMWithTextFilter.png)
+
+In Spring context, the @Id attribute information are important and we should prefer showing those attributes in the Outline:
+
+![Outline of Spring beans with @id filter](images/Symbols/SymbolsBeansWithIdAttrFilter.png)
+
+In other words, displaying attributes or text nodes depends on the XML file.
+
+`Symbols filters` gives the capability to define filter (to include or exclude) some attributes, some text nodes for a given XML file kind.
```
Comment (rgrunber) [inline docs/Symbols.md:23]: This should either be 'Symbol filters' without the back ticks, or `xml.symbols.filters` . Usually things in back ticks indicate they are a key word but in this case "Symbol filters" is just the formal name for the feature.

File: docs/Symbols.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+Symbols filter are composed with:
+
+  * `pattern` (required) : it defines which XML files must use the filter.
```
Comment (rgrunber) [inline docs/Symbols.md:27]: I would say :

a regular expression matching the file names to which this filter should apply.

File: docs/Symbols.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
+# Symbols
+
+## Symbols filter settings
```
Comment (rgrunber) [inline docs/Symbols.md:3]: I have a slight preference for 'Symbol filters'

File: docs/Symbols.md (hunk 1 above)
Comment (angelozerr) [reply docs/Symbols.md]: Fixed thanks for your suggestion.

File: docs/Symbols.md (hunk 2 above)
Comment (angelozerr) [reply docs/Symbols.md]: I agree, I replaced with  `xml.symbols.filters`

File: docs/Symbols.md (hunk 3 above)
Comment (angelozerr) [reply docs/Symbols.md]: fixed

File: docs/Symbols.md (hunk 4 above)
Comment (angelozerr) [reply docs/Symbols.md]: To be consistent with Formatting, I replaced with  `xml.symbols.filters`

File: docs/Symbols.md [hunk 5]
```diff
@@ -10,4 +10,125 @@ The following are also listed:
...
+
+Here a sample Outline with maven `pom.xml`:
+
+![Outline of pom.xml without filter](images/Symbols/SymbolsPOMWithoutFilter.png)
+
+In maven context, the text nodes information are important and we should prefer showing text nodes in the Outline:
+
+![Outline of pom.xml with text filter](images/Symbols/SymbolsPOMWithTextFilter.png)
+
+In Spring context, the @Id attribute information are important and we should prefer showing those attributes in the Outline:
```
Comment (datho7561) [inline docs/Symbols.md:29]: Use backticks for `@Id`

File: docs/Symbols.md [hunk 6]
```diff
@@ -10,4 +10,125 @@ The following are also listed:
...
+
+In other words, displaying attributes or text nodes depends on the XML file.
+
+`xml.symbols.filters` gives the capability to define filter (to include or exclude) some attributes, some text nodes for a given XML file kind.
+
+Symbols filter are composed with:
+
+  * `pattern` (required) : a regular expression matching the file names to which this filter should apply.
+  * `expressions` (required) : defines a list of expression. An expression is composed by:
+    **  `xpath` (required) : defines a basic xpath to declare the attribute, the text node which is concerned by the expression.
```
Comment (datho7561) [inline docs/Symbols.md:41]: In order to get the second level of indent:
```md
 * Level 1 of indent
    * Level 2 of indent
```

File: package.json [hunk 7]
```diff
@@ -356,7 +356,44 @@
...
+          "type": "array",
+          "default": [],
+          "items": {
+            "type": "object",
+            "properties": {
+              "pattern": {
+                "type": "string",
+                "markdownDescription": "File glob pattern. Example: `**/*.Format.ps1xml`\n\nMore information on the glob syntax: https://docs.oracle.com/javase/tutorial/essential/io/fileOps.html#glob"
+              },
+              "expressions": {
```
Comment (datho7561) [inline package.json:372]: Include a markdownDescription for the array, maybe something along the lines of "list of symbol filters to apply. Learn more here"

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: docs/Symbols.md (hunk 5 above)
Comment (angelozerr) [reply docs/Symbols.md]: fixed

File: docs/Symbols.md (hunk 6 above)
Comment (angelozerr) [reply docs/Symbols.md]: fixed

File: docs/Symbols.md [hunk 8]
```diff
@@ -10,4 +10,125 @@ The following are also listed:
...
+            "exclude": "true"
+         },
+         // Include other attributes
+         {
+            "xpath": "//@*",
+         }
+      ]
+   }
+]
+```
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply docs/Symbols.md:133]: done

File: package.json (hunk 7 above)
Comment (angelozerr) [reply package.json:372]: I have add a markdownDescription  for expressions. But the link to the documentation os for xml.symbols.filters which already exists.

File: docs/Symbols.md [hunk 9]
```diff
@@ -1,5 +1,9 @@
 # Symbols
 
+## xml.symbols.enabled
+
+Enable/disable document symbols (Outline). Default is `true`. No symbols are given if `\"xml.symbol.enabled\": false`.
```
Comment (datho7561) [inline docs/Symbols.md:5]: At the end, the setting should be `"xml.symbols.enabled"` (plural)

File: docs/Symbols.md [hunk 10]
```diff
@@ -22,5 +26,144 @@ The following are also listed:
...
+
+## xml.symbols.filters
+
+By default Outline (symbols) display `DOM elements`, `processing instruction` and `DTD element, entity, attribute list declarations`.
+
+The DOM attributes and text nodes are not displayed in the outline. Indeed to ensure good performance, symbols must be limited; displaying attributes and text nodes in the outline could generate a lot of symbols.
+
+Here a sample Outline with maven `pom.xml`:
+
+![Outline of pom.xml without filter](images/Symbols/SymbolsPOMWithoutFilter.png)
```
Comment (datho7561) [inline docs/Symbols.md:40]: The images have white borders, maybe try to crop them?

**Review by datho7561 — CHANGES_REQUESTED** [review]

File: docs/Symbols.md (hunk 9 above)
Comment (angelozerr) [reply docs/Symbols.md]: fixed

File: docs/Symbols.md (hunk 10 above)
Comment (angelozerr) [reply docs/Symbols.md:40]: fixed

Comment (datho7561) [reply docs/Symbols.md:40]: They still have them: 
Screenshot from 2020-10-27 13-50-46

Comment (angelozerr) [reply docs/Symbols.md:40]: Is it better now? I have update the show referenced grammar too to add red square. If you don't like it, please tell me.

Comment (datho7561) [reply docs/Symbols.md:40]: Looks good!

**Review by datho7561 — APPROVED** [review]

**Pull request merged**